### PR TITLE
Add programmatic API

### PR DESCRIPTION
**Pull request by abravalheri**
*Issue number of the reported bug or feature request: #3*

This PR attempts to refactor a bit the existing code so that a Python API is exposed (and `pyproject-fmt` can be used as a library).

**Describe your changes**
- The existing `PyProjectFmtNamespace` class was separated in 2 different classes:
  - `PyProjectFmtCliNamespace` that is used only at the CLI layer and
  - (the new) `PyProjectFmtNamespace`, that is used by all internal functions (I kept the same name here to minimise the size of the changes).
- The signature of `format_pyproject` was changed, allowing it to be easier to use programmatically. Instead of requiring a path to the `pyproject.toml` file, now it allows the contents of the TOML file to be passed directly as a string. 

**Testing performed**
The tests were adjusted accordingly to use the changed Python API.

**Additional context**
In #3 the project author asked about adding documentation via Sphinx.
This change is not added to this PR directly for the sake of keeping the diff small and easier to review. However a follow up change that adds a minimal skeleton for using Sphinx to generate documentation was already worked on, and is available at: https://github.com/abravalheri/pyproject-fmt/pull/1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,3 +33,27 @@ Consult the help for the latest usage:
 ```console
 $ pyproject-fmt --help
 ```
+
+## Programmatic API
```
Comment (gaborbernat) [inline README.md:37]: Oh, please not this. I want sphinx + readthedocs, not an easy-to-go out-of-date markdown blob.

File: src/pyproject_fmt/__main__.py [hunk 2]
```diff
@@ -24,15 +24,15 @@ def color_diff(diff: Iterable[str]) -> Iterable[str]:
 
 
 def run(args: Sequence[str] | None = None) -> int:
-    opts = cli_args(sys.argv[1:] if args is None else args)
-    formatted = format_pyproject(opts)
-    toml = opts.pyproject_toml
-    before = toml.read_text()
+    config = cli_args(sys.argv[1:] if args is None else args)
```
Comment (gaborbernat) [inline src/pyproject_fmt/__main__.py:27]: let's not rename variables in this PR

Comment (gaborbernat) [reply src/pyproject_fmt/__main__.py:27]: Also, I see no reason to change this file part of this PR.

File: src/pyproject_fmt/cli.py [hunk 3]
```diff
@@ -5,13 +5,26 @@
 from pathlib import Path
 from typing import Sequence
 
+DEFAULT_INDENT = 2
```
Comment (gaborbernat) [inline src/pyproject_fmt/cli.py:8]: Unrelated change.

**Review by gaborbernat — CHANGES_REQUESTED** [review]
You've done way too much in this PR that's entirely out of scope here

File: README.md (hunk 1 above)
Comment (abravalheri) [reply README.md:37]: Thank you very much for the review @gaborbernat, I am working on the sphinx + readthedocs separatedly in https://github.com/abravalheri/pyproject-fmt/pull/1.

I am assuming that with this comment you prefer that a quick introduction is not added to the README, so I will go ahead and more this example to the function's docstring.

Comment (gaborbernat) [reply README.md:37]: > I am working on the sphinx + readthedocs separatedly in [abravalheri#1](https://github.com/abravalheri/pyproject-fmt/pull/1).

I'll take this, because I have a peculiar way of doing things :blush: so I'd not burden you with a lot of back and forth to get there.

Comment (gaborbernat) [reply README.md:37]: > I am assuming that with this comment you prefer that a quick introduction is not added to the README, so I will go ahead and more this example to the function's docstring.

I want sphinx documentation generated. This means the sphinx PR needs to go first in before this PR lands.

Comment (abravalheri) [reply README.md:37]: Thank you @gaborbernat, I will wait until your PR lands and then touch this one again.

**Comment by abravalheri** [comment]
Thank you very much for the code review Bernát.

> You've done way too much in this PR that's entirely out of scope here

I am trying to understand exactly what you mean here. Other than variable renames and introducing a constant (that can be easily reverted), would this comment be related to the fact I am splitting `PyProjectFmtCliNamespace` and `PyProjectFmtNamespace`?

The reason why I did that is because I believe for a programmatic use, it is better to have the formatting function accepting an string instead of a file path (the file might simply not exist in the file system, that is my use case in `ini2toml`). If the file exist then API consumers can always read its contents before passing to the API...

Anyway, please let me know what are your concerns and the way you want to go forward with it and I will update the PR accordingly.

**Comment by gaborbernat** [comment]
Yeah, what makes me a bit uncomfortable is the duality of the  PyProjectFmtCliNamespace and PyProjectFmtNamespace :thinking: I need to think what's a better solution :thinking:

File: src/pyproject_fmt/__main__.py [hunk 4]
```diff
@@ -25,14 +25,14 @@ def color_diff(diff: Iterable[str]) -> Iterable[str]:
...
     toml = opts.pyproject_toml
-    before = toml.read_text()
+    before = toml.read_text(encoding="utf-8")
+    formatted = format_pyproject(before, opts)
     changed = before != formatted
     if opts.stdout:  # stdout just prints new format to stdout
         print(formatted, end="")
     else:
-        toml.write_text(formatted)
+        toml.write_text(formatted, encoding="utf-8")
```
Comment (abravalheri) [inline src/pyproject_fmt/__main__.py:35]: I explicitly added encoding (to both read and write) because the [TOML spec requires UTF-8](https://toml.io/en/v1.0.0#spec), but according to the [Python docs](https://docs.python.org/3/library/functions.html#open) the default encoding is [platform dependent](https://docs.python.org/3/library/locale.html#locale.getpreferredencoding)

File: src/pyproject_fmt/formatter/__init__.py [hunk 5]
```diff
@@ -13,8 +13,23 @@ def _perform(parsed: TOMLDocument, opts: PyProjectFmtNamespace) -> None:
...
+    r"""
+    Format a string with the equivalent content of a ``pyproject.toml`` file.
+
+    Usage:
+
+    >>> from pyproject_fmt import PyProjectFmtNamespace, format_pyproject
+    >>> text = '[project]\nname = "myproj"\nversion = "0.1.1"'
+    >>> opts = PyProjectFmtNamespace(indent=2)
+    >>> format_pyproject(text, opts)
+    '[project]\nname = "myproj"\nversion = "0.1.1"\n'
```
Comment (abravalheri) [inline src/pyproject_fmt/formatter/__init__.py:26]: This example here should be kept in sync by the tests (I added `--doctest-modules` to `tox.ini`).

File: tox.ini [hunk 6]
```diff
@@ -22,6 +22,7 @@ extras =
     test
 commands =
     python -m pytest {tty:--color=yes} {posargs: \
+      --doctest-modules \
```
Comment (abravalheri) [inline tox.ini:25]: This was added to make sure the API example is always valid.

File: docs/index.rst [hunk 7]
```diff
@@ -41,3 +41,11 @@ Command line interface
 
    self
    changelog
+
+Python API
```
Comment (abravalheri) [inline docs/index.rst:45]: Here I took the same approach I saw for the documented CLI.
Please let me know if you prefer something more verbore/descriptive.

**Review by abravalheri — COMMENTED** [review]
Thank you very much Bernát.
Meanwhile, I have updated the proposed PR with the docs being handled via Sphinx.

I also changed a little bit the relationship between `PyProjectFmtCliNamespace` and `PyProjectFmtNamespace`, maybe that improves the problems with the duality?

Please let me know if you would like things to be implemented differently.

**Comment by gaborbernat** [comment]
Hey @abravalheri thanks for this, I tinkered around myself a bit locally and came up with https://github.com/tox-dev/pyproject-fmt/pull/9 that I'm more comfortable. I'll close this and land that. Thanks for the PR though, much of your work here helped me with my solution.

**Comment by abravalheri** [comment]
No problems @gaborbernat, it was a pleasure.
Thank you very much for working on this :)
